src/aws/flb_aws_util.c

```diff
@@ -322,15 +328,46 @@ struct flb_http_client *request_do(struct flb_aws_client *aws_client,
         flb_warn("[aws_http_client] failed to increase max response buffer size");
     } 
 
-    /* Add AWS Fluent Bit user agent */
-    if (aws_client->extra_user_agent == NULL) {
+    /* 
+    * Set AWS Fluent Bit user agent FLB_AWS_USER_AGENT
+    * This environment variable needs to be set when:
+    * 1. it is not set;
+    * 2. extra_user_agent has been specified by one aws_client and 
+    * its value doesn't match current AWS Fluent Bit user agent
+    */
+    env = aws_client->upstream->config->env;
+    buf = (char *) flb_env_get(env, "FLB_AWS_USER_AGENT");
+    if (buf == NULL || 
+       (aws_client->extra_user_agent != NULL && strcasecmp(buf, aws_client->extra_user_agent) != 0)) {
+        if (aws_client->extra_user_agent != NULL) {
+            user_agent = aws_client->extra_user_agent;
+        } 
+        else if (getenv(AWS_ECS_METADATA_URI) != NULL) {
+            user_agent = "ecs";
+        } 
+        else {
+            buf = (char *) flb_env_get(env, "k8s");
+            if (buf && strcasecmp(buf, "enabled") == 0) {
+                user_agent = "k8s";
+            }
+        } 
+
+        if (user_agent == NULL) {
+            user_agent = AWS_USER_AGENT_NONE;
+        }
+        
+        flb_env_set(env, "FLB_AWS_USER_AGENT", user_agent);
+    }
+    /* Add AWS Fluent Bit user agent header */
+    buf = (char *) flb_env_get(env, "FLB_AWS_USER_AGENT");
+    if (strcasecmp(buf, AWS_USER_AGENT_NONE) == 0) {
         ret = flb_http_add_header(c, "User-Agent", 10,
                                   "aws-fluent-bit-plugin", 21);
     }
     else {
         user_agent_prefix = flb_sds_create_size(64);
         tmp = flb_sds_printf(&user_agent_prefix, "aws-fluent-bit-plugin-%s",
-                             aws_client->extra_user_agent);
```

Comment:
how does the user agent from FLB_AWS_USER_AGENT get set to this var, 

I do not see a line like:

```
aws_client->extra_user_agent = user_agent
```

Reply:
here: https://github.com/fluent/fluent-bit/blob/0cc7354d65be10449d729dce4be4274caed957bf/src/aws/flb_aws_util.c#L342-L343.

Reply:
And I will set it to the env var.

Reply:
OH I see. I think I misunderstood before. So a better logic might be set the `user_agent` to `client->extra_user_agent` right? And we always take `client->extra_user_agent` as `user_agent_prefix` here. Because user agent should be specific to each instance